motoman_driver/MotoPlus/MotionServer.c

```diff
@@ -1183,18 +1183,34 @@ int Ros_MotionServer_InitTrajPointFullEx(CtrlGroup* ctrlGroup, SmBodyJointTrajPt
 	memcpy(jointTrajData.vel, jointTrajDataEx->vel, sizeof(float)*ROS_MAX_JOINT);
 	memcpy(jointTrajData.acc, jointTrajDataEx->acc, sizeof(float)*ROS_MAX_JOINT);
 
-	return Ros_MotionServer_InitTrajPointFull(ctrlGroup, &jointTrajData);
+	return Ros_MotionServer_InitTrajPointFull(controller, ctrlGroup, &jointTrajData);
 }
 
 //-----------------------------------------------------------------------
 // Setup the first point of a trajectory
 //-----------------------------------------------------------------------
-int Ros_MotionServer_InitTrajPointFull(CtrlGroup* ctrlGroup, SmBodyJointTrajPtFull* jointTrajData)
+int Ros_MotionServer_InitTrajPointFull(Controller* controller, CtrlGroup* ctrlGroup, SmBodyJointTrajPtFull* jointTrajData)
 {
 	long pulsePos[MAX_PULSE_AXES];
 	long curPos[MAX_PULSE_AXES];
 	int i;
 
+    //It's possible for energy-saving mode to activate between /robot_enable and the first trajectory point.
+    //If eco-mode is active, then automatically invoke a /robot_enable.
+    if (Ros_Controller_IsEcoMode(controller))
+    {
+        printf("Energy Saving Function is active. Clearing with automatic robot_enable.\n");
+
+        Ros_MotionServer_StopTrajMode(controller);
+        Ros_Sleep(100); //give time for Ros_Controller_StatusUpdate on other task
+        Ros_MotionServer_StartTrajMode(controller);
+
+        //Give time for the controller to recognize that the INFORM cursor is sitting on
+        //a WAIT instructions. Without this delay, the first call to mpExRcsIncrementMove
+        //will fail with a (-1).
+        Ros_Sleep(100);
+    }
+

```

Comment:
Would there be a more appropriate place to add this functionality?

`Ros_MotionServer_InitTrajPointFull(..)` as a function deals with initialising a nr of data structures.

Making it call `Ros_MotionServer_StartTrajMode(..)` seems out of place here and should not be its responsibility.

Reply:
It was a convenient place to put it, since all code funneled to this function. Is [e946725](https://github.com/ros-industrial/motoman/pull/520/commits/e9467255f207e790dd320b3f680dc7a6f4274d11) more appropriate?

Reply:
Just noticed that this is still open. Would you mind looking at the latest commit?